blackboxopt/optimizers/botorch_base.py

```diff
@@ -44,6 +45,21 @@
     ) from e
 
 
+def _get_numerical_points_from_discrete_space(space: ParameterSpace) -> np.ndarray:
+    """Retrieve all points from a discrete space in the numerical representation"""
+    points_along_dimensions = []
+    for parameter_name in space.get_parameter_names():
+        num_values = space.get_parameter_by_name(parameter_name)[
+            "parameter"
+        ].num_values  # type:ignore
+        points_along_dimensions.append(
+            np.linspace(1 / (2 * num_values), 1 - 1 / (2 * num_values), num_values)
+        )
```

Comment:
This line seems to rely on the internal implementation details of parameterspace and duplicate instead of reuse, also limiting the flexibility to specify custom transformations in the given parameterspace. I'm wondering if such a more explicit alternative that relies on the public parameterspace interface is better suited:
```python
parameter = space.get_parameter_by_name(parameter_name)["parameter"]
if isinstance(parameter, ps.IntegerParameter):
    points_along_dimensions.apend([parameter.val2num(v) for v in range(*parameter.bounds)])
elif isinstance(parameter, ps.OrdinalParameter) or isinstance(parameter, ps.CategoricalParameter):
    points_along_dimensions.append([parameter.val2num(v) for v in parameter.values])
else:
    raise ValueError(f"Only discrete parameters are allowed but got {parameter}")
```

Reply:
Excellent point. I expected that part of the code could be reused but was not sufficiently familiar with the internal workings of `parameterspace`. I implemented your suggestion with a small difference since `range(*parameter.bounds)` needs to include the last element.
```python
        if isinstance(parameter, ps.IntegerParameter):
            bounds = (parameter.bounds[0], parameter.bounds[1] + 1)
            points_along_dimensions.append([parameter.val2num(v) for v in range(*bounds)])
        elif isinstance(parameter, ps.OrdinalParameter) or isinstance(parameter, ps.CategoricalParameter):
            points_along_dimensions.append([parameter.val2num(v) for v in parameter.values])
        else:
            raise ValueError(f"Only discrete parameters are allowed but got {parameter}")
```